File: efficient_go/parsec.go
package efficient

func parserReturn[T any](value T) []T {
    return []T{value}
}

func parserSatisfy[T any](predicate func(T) bool) func([]T) (T, []T) {
    return func(input []T) (T, []T) {
        if len(input) == 0 {
            var defaultValue T
            return defaultValue, input
        }

        if predicate(input[0]) {
            return input[0], input[1:]
        } else {
            var defaultValue T
            return defaultValue, input
        }
    }
}

File: rust/src/parsec.rs
pub fn parser_return<T: Clone>(value: T) -> Result<T, String> {
    Ok(value)
}

pub fn parser_satisfy<T, P>(predicate: P) -> impl Fn(&[T]) -> Result<T, String>
where
    T: Clone,
    P: Fn(&T) -> bool,
{
    move |input: &[T]| {
        if let Some(head) = input.get(0) {
            if predicate(head) {
                Ok(head.clone())
            } else {
                Err("Predicate not satisfied".to_string())
            }
        } else {
            Err("No input".to_string())
        }
    }
}